src/encoder/encoderopus.cpp

```diff
@@ -0,0 +1,439 @@
+    // - Pre-skip: 2 bytes
+    // - Samplerate: 4 bytes
+    // - Output Gain: 2 bytes
+    // - Mapping family: 1 byte
+    // - Channel mapping table: ignored
+    // Total: 19 bytes
+    const int frameSize = 19;
+    unsigned char* data = new unsigned char[frameSize]();
+    int pos = 0; // Current position
+
+    // Magic signature (8 bytes)
+    memcpy(data + pos, "OpusHead", 8);
+    pos += 8;
+
+    // Version number (1 byte, fixed to 1)
+    data[pos++] = 0x01;
+
+    // Channel count (1 byte)
+    data[pos++] = (char)m_channels;
+
+    // Pre-skip (2 bytes, little-endian)
+    int preskip = 0;
+    opus_encoder_ctl(m_pOpus, OPUS_GET_LOOKAHEAD(&preskip));
+    for (int x = 0; x < 2; x++) {
+        data[pos++] = (preskip >> (x*8)) & 0xFF;
+    }
+
+    // Sample rate (4 bytes, little endian)
+    for (int x = 0; x < 4; x++) {
+        data[pos++] = (m_samplerate >> (x*8)) & 0xFF;
+    }
+
+    // Output gain (2 bytes, little-endian, fixed to 0)
+    data[pos++] = 0;
+    data[pos++] = 0;
+
+    // Channel mapping (1 byte, fixed to 0, means one stream)
+    data[pos++] = 0;
+
+    // Ignore channel mapping table
+
+    // Push finished header to stream
+    ogg_packet packet;
+    packet.b_o_s = 1;
+    packet.e_o_s = 0;
+    packet.granulepos = 0;
+    packet.packetno = m_packetNumber++;
+    packet.packet = data;
+    packet.bytes = frameSize;
+
+    ogg_stream_packetin(&m_oggStream, &packet);
+    delete[] data;
+}
+
+void EncoderOpus::pushTagsPacket() {
+    // Opus comment header
+    // Format from https://tools.ietf.org/html/rfc7845.html#section-5.2
+
+    QByteArray combinedComments;
+    int commentCount = 0;
+
+    const char* vendorString = opus_get_version_string();
+    int vendorStringLength = strlen(vendorString);
+
+    // == Compute tags frame size ==
+    // - Magic signature: 8 bytes
+    // - Vendor string length: 4 bytes
+    // - Vendor string: dynamic size
+    // - Comment list length: 4 bytes
+    int frameSize = 8 + 4 + vendorStringLength + 4;
+    // - Comment list: dynamic size
+    QMapIterator<QString, QString> iter(m_opusComments);
+    while(iter.hasNext()) {
+        iter.next();
+        QString comment = iter.key() + "=" + iter.value();
+        QByteArray commentBytes = comment.toUtf8();
+        int commentBytesLength = commentBytes.size();
+
+        // One comment is:
+        // - 4 bytes of string length
+        // - string data
+
+        // Add comment length field and data to comments "list"
+        for (int x = 0; x < 4; x++) {
+            unsigned char fieldValue = (commentBytesLength >> (x*8)) & 0xFF;
+            combinedComments.append(fieldValue);
+        }
+        combinedComments.append(commentBytes);
+
+        // Don't forget to include this comment in the overall size calculation
+        frameSize += (4 + commentBytesLength);
+        commentCount++;
+    }
+
+    // == Actual frame building ==
+    unsigned char* data = new unsigned char[frameSize]();
```

Comment:
Please avoid old skool manual memory management and instead use std::vector or QVector.